libs/ledger-live-common/src/families/vechain/js-transaction.ts

```diff
@@ -0,0 +1,135 @@
+import { Account, TransactionCommon } from "@ledgerhq/types-live";
+import BigNumber from "bignumber.js";
+import { $Shape } from "utility-types";
+import { DEFAULT_GAS_COEFFICIENT, HEX_PREFIX, MAINNET_CHAIN_TAG } from "./constants";
+import { Transaction } from "./types";
+import { Transaction as ThorTransaction } from "thor-devkit";
+import { calculateFee, estimateGas, generateNonce, getBlockRef } from "./utils/transaction-utils";
+import { VTHO_ADDRESS } from "./contracts/constants";
+import VIP180 from "./contracts/abis/VIP180";
+import { calculateTransactionInfo } from "./utils/calculateTransactionInfo";
+import { isValid } from "./utils/address-utils";
+
+/**
+ * Create an empty VET or VTHO transaction
+ *
+ * @returns {Transaction}
+ */
+export const createTransaction = (): Transaction => {
+  const chainTag = MAINNET_CHAIN_TAG;
+
+  return {
+    family: "vechain",
+    body: {
+      chainTag,
+      // placeholder, if "empty" returns error on send modal open
+      blockRef: "0x0000000000000000",
+      expiration: 18,
+      clauses: [],
+      gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
+      gas: "0",
+      dependsOn: null,
+      nonce: generateNonce(),
+    },
+    amount: BigNumber(0),
+    estimatedFees: BigNumber(0),
+    recipient: "",
+    useAllAmount: false,
+    networkInfo: true,
+  };
+};
+
+/**
+ * Apply patch to a transaction
+ *
+ * @param {Transaction} t
+ * @param {TransactionCommon} patch
+ * @returns patched transaction
+ */
+export const updateTransaction = (
+  t: Transaction,
+  patch: $Shape<TransactionCommon>,
+): Transaction => {
+  return { ...t, ...patch };
+};
+
+/**
+ * Prepare transaction before checking status
+ *
+ * @param {Account} a
+ * @param {Transaction} t
+ */
+export const prepareTransaction = async (
+  account: Account,
+  transaction: Transaction,
+): Promise<Transaction> => {
+  const { amount, isTokenAccount } = await calculateTransactionInfo(account, transaction);
+
+  let blockRef = "";
+
+  let clauses: Array<ThorTransaction.Clause> = [];
+  if (transaction.recipient && isValid(transaction.recipient)) {
+    blockRef = await getBlockRef();
+    if (isTokenAccount) {
+      clauses = await calculateClausesVtho(transaction, amount);
+    } else {
+      clauses = await calculateClausesVet(transaction, amount);
+    }
+  }
+
+  const gas = await estimateGas({
+    ...transaction,
+    body: { ...transaction.body, clauses: clauses },
+  });
+
+  const estimatedFees = await calculateFee(new BigNumber(gas), transaction.body.gasPriceCoef);
+
+  const body = { ...transaction.body, gas, blockRef, clauses };
+
+  return { ...transaction, body, amount, estimatedFees };
+};
+
+export const calculateClausesVtho = async (
```

Comment:
Unfortunately, it needs it because it returns an array of promises and ts raises an error:

![Screenshot 2023-09-11 at 01 12 53](https://github.com/LedgerHQ/ledger-live/assets/5785838/ca4d6687-d06e-4f67-97fb-d5f9702ac9fb)